### PR TITLE
CORE-19572 Robust message processing

### PR DESCRIPTION
**Pull request by simon-johnson-r3**
In the JIRA ticket we see bad permissions set up for a vnode crypto db. The code assumed all tenantIds had accessible crypto dbs and threw an exception if not. In the case of bad user permissions (possible for example if bringing your own vnode db) this is not the case, so we need to be more robust.

Catch exceptions and ignore where possible to enable as much of key rotation as possible to continue.

### CODE REVIEW

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt [hunk 1]
```diff
@@ -64,103 +64,119 @@ class CryptoRekeyBusProcessor(
         logger.debug("received ${events.size} key rotation requests")
 
         events.mapNotNull { it.timestamp to it.value }.forEach { (timestamp, request) ->
-            logger.debug("processing $request")
-            require(request != null)
-
-            if (!hasPreviousRotationFinished()) {
-                logger.info("A key rotation is already ongoing, ignoring request to start new one.")
-                return emptyList()
+            try {
```
Comment (simon-johnson-r3) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:67]: the github diff is not great here, but in both message handlers I've moved the entire processing into its own function which is wrapped in a try/catch here. This is not directly related to the original bug, but we shouldn't fail all messages in this poll simply because one throws an exception.

File: components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt [hunk 2]
```diff
@@ -64,103 +64,119 @@ class CryptoRekeyBusProcessor(
...
+                val virtualNodeTenantIds = virtualNodeInfo.map { it.holdingIdentity.shortHash.toString() }
+
+                // We do not need to use separate wrapping repositories for the different cluster level tenants,
+                // since they share the cluster crypto database. So we scan over the virtual node tenants and an arbitrary
+                // choice of cluster level tenant. We pick CryptoTenants.CRYPTO as the arbitrary cluster level tenant,
+                // and we should not also check CryptoTenants.P2P and CryptoTenants.REST since if we do we'll get duplicate.
+                val allTenantIds = virtualNodeTenantIds + listOf(CryptoTenants.CRYPTO)
+                logger.debug("Found ${allTenantIds.size} tenants; first few are: ${allTenantIds.take(10)}")
+                val targetWrappingKeys = allTenantIds.asSequence().map { tenantId ->
+                    try {
```
Comment (simon-johnson-r3) [inline components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:118]: This try/catch is the bug fix.

File: components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessorTests.kt [hunk 3]
```diff
@@ -214,6 +215,37 @@ class CryptoRekeyBusProcessorTests {
...
+        )
+        whenever(wrappingRepositoryFactory.create(any())).thenReturn(wrappingRepository)
+
+        cryptoRekeyBusProcessor.onNext(listOf(getUnmanagedKeyRotationKafkaRecord()))
+
+        verify(rewrapPublisher, times(1)).publish(any())
+        assertThat(rewrapPublishCapture.allValues).hasSize(1)
+
+        val allTenants = virtualNodeTenantIds + CryptoTenants.CRYPTO
+        assertThat(rewrapPublishCapture.firstValue).hasSize(allTenants.size)
```
Comment (anien) [inline components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessorTests.kt]: I am a bit confused. Do we expect to publish for all tenants even when there was some error in between? I expected seeing `assertThat(rewrapPublishCapture.firstValue).hasSize(allTenants.size - 1)`

Comment (simon-johnson-r3) [reply components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessorTests.kt]: well spotted!

this is a really subtle effect of using sequences in the implementation, they are resolved at the point of use not creation, so what actually happens here is the `writeStateForUnmanagedKey` function triggers the exception the first time the sequence is "pulled" (the test fails if you remove the exception handling for this reason), but then when the same sequence is "pulled" again in `publishIndividualUnmanagedRewrappingRequests` all the iterations return valid wrapping key info instances, and there is no exception

it's an interaction problem relating to the use of a mock which behaves differently for the same input when called more than once, however it does illustrate the use of sequences is quite inefficient, so I'll remove that from the implementation

**Review by anien — DISMISSED** [review]
LGTM.

**Review by anien — APPROVED** [review]
